WordPressKitTests/AllDomainsResultDomainTests.swift

```diff
@@ -0,0 +1,163 @@
+import Foundation
+import XCTest
+
+@testable import WordPressKit
+
+final class AllDomainsResultDomainTests: XCTestCase {
+
+    // MARK: - Properties
+
+    private let dateFormatter = ISO8601DateFormatter()
+
+    // MARK: - Tests
+
+    /// Tests decoding with default input defined in `Defaults` enum.
+    func testDecodingWithDefaultInput() throws {
+        // Given
+        let decoder = makeDecoder()
+        let input = try makeInput()
+
+        // When
+        let output = try decoder.decode(Domain.self, from: input)
+
+        // Then
+        let expectedOutput = makeDomain()
+        assertEqual(output, otherDomain: expectedOutput)
+    }
+
+    /// Tests decoding with empty `registrationDate` and `expiryDate`.
+    func testDecodingWithEmptyExpiryAndRegistrationDates() throws {
+        // Given
+        let decoder = makeDecoder()
+        let input = try makeInput(registrationDate: "", expiryDate: nil)
+
+        // When
+        let output = try decoder.decode(Domain.self, from: input)
+
+        // Then
+        let expectedOutput = makeDomain(registrationDate: "", expiryDate: nil)
+        assertEqual(output, otherDomain: expectedOutput)
+    }
+
+    /// Tests decoding with `type = mapping` and `hasRegistration = true`.
+    func testDecodingWithRegistrationAndMappingType() throws {
+        // Given
+        let decoder = makeDecoder()
+        let input = try makeInput(type: "mapping", hasRegistration: true)
+
+        // When
+        let output = try decoder.decode(Domain.self, from: input)
+
+        // Then
+        let expectedOutput = makeDomain(type: .registered, hasRegistration: true)
+        assertEqual(output, otherDomain: expectedOutput)
+    }
+
+    // MARK: - Helpers
+
+    private func assertEqual(_ domain: Domain, otherDomain: Domain) {
+        XCTAssertEqual(domain.domain, otherDomain.domain)
+        XCTAssertEqual(domain.blogId, otherDomain.blogId)
+        XCTAssertEqual(domain.blogName, otherDomain.blogName)
+        XCTAssertEqual(domain.type, otherDomain.type)
+        XCTAssertEqual(domain.isDomainOnlySite, otherDomain.isDomainOnlySite)
+        XCTAssertEqual(domain.isWpcomStagingDomain, otherDomain.isWpcomStagingDomain)
+        XCTAssertEqual(domain.registrationDate, otherDomain.registrationDate)
+        XCTAssertEqual(domain.expiryDate, otherDomain.expiryDate)
+        XCTAssertEqual(domain.wpcomDomain, otherDomain.wpcomDomain)
+        XCTAssertEqual(domain.currentUserIsOwner, otherDomain.currentUserIsOwner)
+        XCTAssertEqual(domain.siteSlug, otherDomain.siteSlug)
+        XCTAssertEqual(domain.status?.value, otherDomain.status?.value)
+        XCTAssertEqual(domain.status?.type, otherDomain.status?.type)
+    }
```

Comment:
Have you considered making `Domain`, aka `DomainsServiceRemote.AllDomainsListItem`, `Equatable`?

Reply:
My thinking is that if a logic is only needed for the purpose of testing, then that logic should only exist in the test environment.

Reply:
> I believe that if a specific logic or behavior is solely required for testing purposes, it should be confined to the test environment.

Agreed. At the same time, if we compromised and defined `Equatable` in the production code, this code would simplify to


```swift
extension DomainsServiceRemote.AllDomainsListItem: Equatable {}
extension DomainsServiceRemote.AllDomainsListItem.Status: Equatable {}

XCTAssertEqual(output, expectedOutput)
```

With the added benefit that the compiler would update the check for us if a property is added, removed, or changes type.

Seems like a worthy trade to me.

Reply:
> that deep comparison logic is only needed in a test environment. In a non-test environment, usually these models are considered equal if their primary attributes are equal

Interesting... 🤔 Is that justified? I can see how `a.domain != b.domain` obviously means that the `Domain` are different. But what about `a.domain == b.domain; a.isWpcomStagingDomain != b.isWpcomStagingDomain`? That should count as a different domain for me. 

I can see the argument "in practice that will never happen," but that's a weak assumption to make. If the domain modeling allows it to happen, then we I think ought to handle it.

That's a broader conversation than this PR, though.